log-service/src/main/kotlin/org/vivlaniv/nexohub/Application.kt

```diff
@@ -94,36 +99,39 @@ fun main() {
         log.info("Batch of $writtenRows log messages inserted")
     }
 
+    launch(Dispatchers.IO) {
+        while (true) {
+            delay(30_000)
+            batch.flush()
+        }
+    }
+
     // subscribe on topic
     redissonClient.getTopic(logsTopic).addListener(String::class.java) { _, msg ->
-        batch.add(Json.decodeFromString(msg))
+        launch {
+            batch.add(Json.decodeFromString(msg))
+        }
     }
 
     log.info("log-service started")
 }
 
 class Batch<T>(private val size: Int = 500_000, private val action: (List<T>) -> Unit) {
-    private var batch: MutableList<T>
-
-    init {
-        batch = ArrayList(size)
-        Thread{
-            fixedRateTimer(period = 30_000) { flush() }
-        }.start()
-    }
+    private val mutex = Mutex()
+    private var batch: MutableList<T> = ArrayList(size)
 
-    private fun flush() {
+    suspend fun flush() {
         if (batch.isEmpty()) return
-        val oldBatch = synchronized(this) {
+        val oldBatch = mutex.withLock {
             val oldBatch = batch
             batch = ArrayList(size)
             oldBatch
         }
```

Comment:
Also it can be done like this:

```kotlin
        val oldBatch = {
          newBatch = ArrayList(size)
          mutex.withLock {
            batch.also { batch = newBatch }
          }
        } 
```

Reply:
Так в котлине делают свапы видимо) https://stackoverflow.com/questions/45377802/swap-function-in-kotlin